File: 881-boatsToSavePeople/numRescueBoats.ts
function numRescueBoats(people: number[], limit: number): number {
    people.sort((a, b) => a - b);
    let numBoats: number = 0;
    let bigPtr: number = people.length - 1;
    let smallPtr: number = 0;
    while (smallPtr <= bigPtr) {
        if (people[bigPtr] + people[smallPtr] <= limit) {
            smallPtr++;
        }
        bigPtr--;
        numBoats++;
    }
    return numBoats;
};

// Tests
console.log(numRescueBoats([1, 2], 3)); // 1
console.log(numRescueBoats([3, 2, 2, 1], 3)); // 3
console.log(numRescueBoats([3, 5, 3, 4], 5)); // 4

File: 881-boatsToSavePeople/numRescueBoats.go
package main

import (
	"slices"
)

func numRescueBoats(people []int, limit int) int {
	slices.Sort(people)
	smallestIdx, biggestIdx := 0, len(people)-1
	totalBoats := 0
	for smallestIdx <= biggestIdx {
		if people[smallestIdx]+people[biggestIdx] <= limit {
			smallestIdx++
		}
		biggestIdx--
		totalBoats++
	}
	return totalBoats
}
